File: python/networking/mail_checker.py
"""Mail Checker — monitor an inbox for new messages with filtering and notification."""

from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Optional, Callable
import hashlib


@dataclass
class EmailAddress:
    """A parsed email address."""
    name: str = ""
    address: str = ""

    def __str__(self) -> str:
        if self.name:
            return f"{self.name} <{self.address}>"
        return self.address

    @staticmethod
    def parse(raw: str) -> "EmailAddress":
        raw = raw.strip()
        if "<" in raw and ">" in raw:
            name = raw[:raw.index("<")].strip().strip('"')
            addr = raw[raw.index("<") + 1:raw.index(">")].strip()
            return EmailAddress(name=name, address=addr)
        return EmailAddress(address=raw)


@dataclass
class MailMessage:
    """A single email message."""
    uid: str
    sender: EmailAddress
    recipients: list[EmailAddress]
    subject: str
    body: str
    date: datetime
    is_read: bool = False
    is_flagged: bool = False
    headers: dict[str, str] = field(default_factory=dict)

    @property
    def preview(self) -> str:
        body_preview = self.body[:80].replace("\n", " ")
        if len(self.body) > 80:
            body_preview += "..."
        return body_preview

    def matches_filter(self, f: "MailFilter") -> bool:
        if f.sender_contains and f.sender_contains.lower() not in self.sender.address.lower():
            return False
        if f.subject_contains and f.subject_contains.lower() not in self.subject.lower():
            return False
        if f.body_contains and f.body_contains.lower() not in self.body.lower():
            return False
        if f.is_unread_only and self.is_read:
            return False
        if f.since and self.date < f.since:
            return False
        return True

    def __repr__(self) -> str:
        flag = "★" if self.is_flagged else " "
        read = " " if self.is_read else "●"
        date_str = self.date.strftime("%Y-%m-%d %H:%M")
        return f"{read}{flag} {date_str} {self.sender.address:30s} {self.subject}"


@dataclass
class MailFilter:
    """Filter criteria for mail searches."""
    sender_contains: str = ""
    subject_contains: str = ""
    body_contains: str = ""
    is_unread_only: bool = False
    since: Optional[datetime] = None


@dataclass
class CheckResult:
    """Result of a mail check operation."""
    total_messages: int
    new_messages: int
    matched_messages: list[MailMessage]
    checked_at: datetime = field(default_factory=lambda: datetime.now(timezone.utc))

    def summary(self) -> str:
        lines = [
            f"Mail check at {self.checked_at.strftime('%Y-%m-%d %H:%M:%S UTC')}",
            f"  Total: {self.total_messages}, New: {self.new_messages}, Matched: {len(self.matched_messages)}",
        ]
        for msg in self.matched_messages[:10]:
            lines.append(f"  {msg}")
        if len(self.matched_messages) > 10:
            lines.append(f"  ... and {len(self.matched_messages) - 10} more")
        return "\n".join(lines)


class Mailbox:
    """An in-process mailbox that stores messages and supports check/filter operations.

    Models the data layer of a mail client — the same operations that IMAP/POP3
    provide over the network. Actual protocol integration is transport-level;
    this module is the callable library underneath.
    """

    def __init__(self, owner: str) -> None:
        self.owner = owner
        self.messages: dict[str, MailMessage] = {}
        self.last_check: Optional[datetime] = None
        self._seen_uids: set[str] = set()
        self._notifications: list[Callable[[MailMessage], None]] = []

    def deliver(self, msg: MailMessage) -> None:
        """Deliver a message to the mailbox (called by the mail server)."""
        self.messages[msg.uid] = msg
        if msg.uid not in self._seen_uids:
            for callback in self._notifications:
                callback(msg)

    def check(self, mail_filter: Optional[MailFilter] = None) -> CheckResult:
        """Check for messages, optionally filtered. Returns a CheckResult."""
        new_uids = set(self.messages.keys()) - self._seen_uids
        self._seen_uids = set(self.messages.keys())
        now = datetime.now(timezone.utc)

        all_msgs = sorted(self.messages.values(), key=lambda m: m.date, reverse=True)

        if mail_filter:
            matched = [m for m in all_msgs if m.matches_filter(mail_filter)]
        else:
            matched = all_msgs

        result = CheckResult(
            total_messages=len(self.messages),
            new_messages=len(new_uids),
            matched_messages=matched,
            checked_at=now,
        )
        self.last_check = now
        return result

    def mark_read(self, uid: str) -> bool:
        """Mark a message as read."""
        if uid in self.messages:
            self.messages[uid].is_read = True
            return True
        return False

    def mark_flagged(self, uid: str, flagged: bool = True) -> bool:
        """Flag or unflag a message."""
        if uid in self.messages:
            self.messages[uid].is_flagged = flagged
            return True
        return False

    def delete(self, uid: str) -> bool:
        """Delete a message from the mailbox."""
        if uid in self.messages:
            del self.messages[uid]
            self._seen_uids.discard(uid)
            return True
        return False

    def on_new_mail(self, callback: Callable[[MailMessage], None]) -> None:
        """Register a callback for new mail notifications."""
        self._notifications.append(callback)

    def unread_count(self) -> int:
        """Return the number of unread messages."""
        return sum(1 for m in self.messages.values() if not m.is_read)

    def stats(self) -> dict:
        """Return mailbox statistics."""
        msgs = list(self.messages.values())
        return {
            "owner": self.owner,
            "total": len(msgs),
            "unread": sum(1 for m in msgs if not m.is_read),
            "flagged": sum(1 for m in msgs if m.is_flagged),
            "latest": max((m.date for m in msgs), default=None),
        }


class MailServer:
    """A simple mail server that routes messages between mailboxes."""

    def __init__(self) -> None:
        self.mailboxes: dict[str, Mailbox] = {}
        self._next_uid = 0

    def create_mailbox(self, owner: str) -> Mailbox:
        """Create a mailbox for a user."""
        mb = Mailbox(owner)
        self.mailboxes[owner] = mb
        return mb

    def send(self, sender: EmailAddress, recipients: list[EmailAddress],
             subject: str, body: str, headers: Optional[dict[str, str]] = None) -> MailMessage:
        """Send a message. Delivers to all recipients with local mailboxes."""
        self._next_uid += 1
        uid = hashlib.sha256(f"{self._next_uid}-{subject}".encode()).hexdigest()[:12]
        msg = MailMessage(
            uid=uid,
            sender=sender,
            recipients=recipients,
            subject=subject,
            body=body,
            date=datetime.now(timezone.utc),
            headers=headers or {},
        )
        for recip in recipients:
            if recip.address in self.mailboxes:
                self.mailboxes[recip.address].deliver(msg)
        return msg

    def get_mailbox(self, owner: str) -> Optional[Mailbox]:
        return self.mailboxes.get(owner)


def poll_check(mailbox: Mailbox, mail_filter: Optional[MailFilter] = None,
               interval_label: str = "") -> CheckResult:
    """Perform a single poll check — the operation that a scheduler would invoke."""
    result = mailbox.check(mail_filter)
    return result


# --- demo ---
if __name__ == "__main__":
    server = MailServer()
    nathan = server.create_mailbox("nathan@em.com")
    kathryn = server.create_mailbox("kathryn@em.com")

    # Kathryn sends Nathan a finance report
    server.send(
        sender=EmailAddress("Kathryn Lyonne", "kathryn@em.com"),
        recipients=[EmailAddress("Nathan", "nathan@em.com")],
        subject="Q2 Finance Positions",
        body="Morning briefing: forex positions are tracking ahead of pace. "
             "Covered Captivate and Anthropic bills. DigitalOcean due in 3 days.",
    )
    server.send(
        sender=EmailAddress("Jay Harper", "jay@em.com"),
        recipients=[EmailAddress("Nathan", "nathan@em.com")],
        subject="Git commit summary",
        body="12 commits across 3 repos. Rosetta Stone: G036-G038 complete.",
    )
    server.send(
        sender=EmailAddress("Spam Bot", "spam@junk.com"),
        recipients=[EmailAddress("Nathan", "nathan@em.com")],
        subject="You won a prize!",
        body="Click here to claim your free vacation.",
    )

    # Check with filter
    result = nathan.check(MailFilter(sender_contains="em.com"))
    print(result.summary())
    print(f"\nUnread: {nathan.unread_count()}")
    print(f"Stats: {nathan.stats()}")

File: go/networking/mail_checker.go
// Mail Checker — monitor an inbox for new messages with filtering and notification.
package networking

import (
	"fmt"
	"sort"
	"strings"
	"sync"
	"time"
)

// MailAddress is a parsed email address.
type MailAddress struct {
	Name    string
	Address string
}

func (a MailAddress) String() string {
	if a.Name != "" {
		return fmt.Sprintf("%s <%s>", a.Name, a.Address)
	}
	return a.Address
}

// ParseMailAddress parses "Name <addr>" or plain "addr".
func ParseMailAddress(raw string) MailAddress {
	raw = strings.TrimSpace(raw)
	if lt := strings.Index(raw, "<"); lt >= 0 {
		if gt := strings.Index(raw, ">"); gt > lt {
			name := strings.TrimSpace(raw[:lt])
			name = strings.Trim(name, "\"")
			addr := strings.TrimSpace(raw[lt+1 : gt])
			return MailAddress{Name: name, Address: addr}
		}
	}
	return MailAddress{Address: raw}
}

// MailMsg is a single email message.
type MailMsg struct {
	UID        string
	Sender     MailAddress
	Recipients []MailAddress
	Subject    string
	Body       string
	Date       time.Time
	IsRead     bool
	IsFlagged  bool
	Headers    map[string]string
}

// Preview returns the first 80 chars of the body.
func (m MailMsg) Preview() string {
	body := strings.ReplaceAll(m.Body, "\n", " ")
	if len(body) > 80 {
		return body[:80] + "..."
	}
	return body
}

// MailFilter defines criteria for filtering messages.
type MailFilter struct {
	SenderContains  string
	SubjectContains string
	BodyContains    string
	IsUnreadOnly    bool
	Since           *time.Time
}

// Matches checks if a message matches the filter.
func (f *MailFilter) Matches(m *MailMsg) bool {
	if f.SenderContains != "" &&
		!strings.Contains(strings.ToLower(m.Sender.Address), strings.ToLower(f.SenderContains)) {
		return false
	}
	if f.SubjectContains != "" &&
		!strings.Contains(strings.ToLower(m.Subject), strings.ToLower(f.SubjectContains)) {
		return false
	}
	if f.BodyContains != "" &&
		!strings.Contains(strings.ToLower(m.Body), strings.ToLower(f.BodyContains)) {
		return false
	}
	if f.IsUnreadOnly && m.IsRead {
		return false
	}
	if f.Since != nil && m.Date.Before(*f.Since) {
		return false
	}
	return true
}

// MailCheckResult holds the result of a check operation.
type MailCheckResult struct {
	TotalMessages   int
	NewMessages     int
	MatchedMessages []MailMsg
	CheckedAt       time.Time
}

// Summary returns a human-readable summary.
func (r MailCheckResult) Summary() string {
	var b strings.Builder
	fmt.Fprintf(&b, "Mail check at %s\n", r.CheckedAt.Format("2006-01-02 15:04:05 UTC"))
	fmt.Fprintf(&b, "  Total: %d, New: %d, Matched: %d\n",
		r.TotalMessages, r.NewMessages, len(r.MatchedMessages))
	limit := 10
	if len(r.MatchedMessages) < limit {
		limit = len(r.MatchedMessages)
	}
	for _, msg := range r.MatchedMessages[:limit] {
		read := "●"
		if msg.IsRead {
			read = " "
		}
		flag := " "
		if msg.IsFlagged {
			flag = "★"
		}
		fmt.Fprintf(&b, "  %s%s %s %-30s %s\n",
			read, flag, msg.Date.Format("2006-01-02 15:04"), msg.Sender.Address, msg.Subject)
	}
	if len(r.MatchedMessages) > 10 {
		fmt.Fprintf(&b, "  ... and %d more\n", len(r.MatchedMessages)-10)
	}
	return b.String()
}

// MailboxStore is an in-process mailbox.
type MailboxStore struct {
	mu        sync.RWMutex
	Owner     string
	messages  map[string]*MailMsg
	seenUIDs  map[string]bool
	LastCheck *time.Time
}

// NewMailboxStore creates a mailbox for the given owner.
func NewMailboxStore(owner string) *MailboxStore {
	return &MailboxStore{
		Owner:    owner,
		messages: make(map[string]*MailMsg),
		seenUIDs: make(map[string]bool),
	}
}

// Deliver adds a message to the mailbox.
func (mb *MailboxStore) Deliver(msg MailMsg) {
	mb.mu.Lock()
	defer mb.mu.Unlock()
	mb.messages[msg.UID] = &msg
}

// Check performs a mail check, returning results with optional filter.
func (mb *MailboxStore) Check(filter *MailFilter) MailCheckResult {
	mb.mu.Lock()
	defer mb.mu.Unlock()

	newCount := 0
	for uid := range mb.messages {
		if !mb.seenUIDs[uid] {
			newCount++
			mb.seenUIDs[uid] = true
		}
	}

	now := time.Now().UTC()
	mb.LastCheck = &now

	msgs := make([]MailMsg, 0, len(mb.messages))
	for _, m := range mb.messages {
		msgs = append(msgs, *m)
	}
	sort.Slice(msgs, func(i, j int) bool { return msgs[i].Date.After(msgs[j].Date) })

	var matched []MailMsg
	if filter != nil {
		for _, m := range msgs {
			if filter.Matches(&m) {
				matched = append(matched, m)
			}
		}
	} else {
		matched = msgs
	}

	return MailCheckResult{
		TotalMessages:   len(mb.messages),
		NewMessages:     newCount,
		MatchedMessages: matched,
		CheckedAt:       now,
	}
}

// MarkRead marks a message as read.
func (mb *MailboxStore) MarkRead(uid string) bool {
	mb.mu.Lock()
	defer mb.mu.Unlock()
	if m, ok := mb.messages[uid]; ok {
		m.IsRead = true
		return true
	}
	return false
}

// MarkFlagged flags or unflags a message.
func (mb *MailboxStore) MarkFlagged(uid string, flagged bool) bool {
	mb.mu.Lock()
	defer mb.mu.Unlock()
	if m, ok := mb.messages[uid]; ok {
		m.IsFlagged = flagged
		return true
	}
	return false
}

// Delete removes a message.
func (mb *MailboxStore) Delete(uid string) bool {
	mb.mu.Lock()
	defer mb.mu.Unlock()
	if _, ok := mb.messages[uid]; ok {
		delete(mb.messages, uid)
		delete(mb.seenUIDs, uid)
		return true
	}
	return false
}

// UnreadCount returns the number of unread messages.
func (mb *MailboxStore) UnreadCount() int {
	mb.mu.RLock()
	defer mb.mu.RUnlock()
	count := 0
	for _, m := range mb.messages {
		if !m.IsRead {
			count++
		}
	}
	return count
}

// Total returns the total number of messages.
func (mb *MailboxStore) Total() int {
	mb.mu.RLock()
	defer mb.mu.RUnlock()
	return len(mb.messages)
}

// MailServerStore routes messages between mailboxes.
type MailServerStore struct {
	mu        sync.Mutex
	mailboxes map[string]*MailboxStore
	nextUID   int
}

// NewMailServerStore creates a mail server.
func NewMailServerStore() *MailServerStore {
	return &MailServerStore{
		mailboxes: make(map[string]*MailboxStore),
	}
}

// CreateMailbox creates a mailbox for the given owner address.
func (s *MailServerStore) CreateMailbox(owner string) *MailboxStore {
	s.mu.Lock()
	defer s.mu.Unlock()
	mb := NewMailboxStore(owner)
	s.mailboxes[owner] = mb
	return mb
}

// Send routes a message to all local recipients.
func (s *MailServerStore) Send(sender MailAddress, recipients []MailAddress,
	subject, body string) string {
	s.mu.Lock()
	s.nextUID++
	uid := fmt.Sprintf("msg-%08x", s.nextUID)
	s.mu.Unlock()

	msg := MailMsg{
		UID:        uid,
		Sender:     sender,
		Recipients: recipients,
		Subject:    subject,
		Body:       body,
		Date:       time.Now().UTC(),
		Headers:    make(map[string]string),
	}
	for _, r := range recipients {
		s.mu.Lock()
		mb, ok := s.mailboxes[r.Address]
		s.mu.Unlock()
		if ok {
			mb.Deliver(msg)
		}
	}
	return uid
}
